WIP AssetGalleryField -> AssetAdmin API refactor

chillu opened the pull request:
See https://github.com/silverstripe/silverstripe-framework/issues/5725 for context. Still requires a few more tests before it's ready to merge.

tractorcow commented on client/src/boot/index.js:
```diff
@@ -1,66 +1,50 @@
+      const createEndpoint = function(endpointConfig) {
+        return backend.createEndpointFetcher(Object.assign(
+            endpointConfig,
+            { defaultData: { SecurityID: Config.get('SecurityID') } }
```
![image](https://cloud.githubusercontent.com/assets/936064/16471853/c173c562-3eb3-11e6-8b05-94542f5b1a14.png)

tractorcow replied:
I think a {} at the start should fix that. :D

tractorcow commented on client/src/boot/index.js:
```diff
@@ -1,66 +1,50 @@
     onadd() {
-      const store = configureStore();
+      const sectionConfig = Config.getSection('SilverStripe\\AssetAdmin\\Controller\\AssetAdmin');
+      const store = configureStore()
```
Semicolon

tractorcow commented on client/src/boot/index.js:
```diff
@@ -1,66 +1,50 @@
 
 $.entwine('ss', () => {
-  $('.asset-gallery-component-wrapper').entwine({
+  $('#assetadmin-cms-content .cms-content-view').entwine({
```
This needs to go all the way up to `#assetadmin-cms-content` otherwise you won't be able to react-control breadcrumbs. I've checked campaign admin and this is the same level that area is rooted.

chillu replied:
Good catch - I'm doing this as part of the react-router refactoring, since react-router needs control over the whole right hand side for all React powered sections. I don't think this blocks the merge of this PR?

tractorcow commented on code/Controller/AssetAdmin.php:
```diff
@@ -78,335 +75,405 @@ class AssetAdmin extends LeftAndMain implements PermissionProvider
+    {
+        parent::init();
+
+        Requirements::javascript(FRAMEWORK_DIR . "/admin/client/dist/js/bundle-lib.js");
```
This file is already included in the `parent::init`, but also has the `provides`, so you don't need the extra bundle-lib here.

tractorcow commented on code/Controller/AssetAdmin.php:
```diff
@@ -78,335 +75,405 @@ class AssetAdmin extends LeftAndMain implements PermissionProvider
+    }
 
 	public function getClientConfig() {
+        $baseLink = $this->Link();
```
Mixed indentation here. :) Set your IDE to respect .editorconfig?

tractorcow replied:
Ok, it looks like the new code = correct, and there is some remaining incorrect indentation. Still, worth a reformat if you can help it.

chillu replied:
Done anyway, had Atom set up with editorconfig, but not PHPStorm :)

flamerohr commented on code/Controller/AssetAdmin.php:
```diff
@@ -78,335 +75,405 @@ class AssetAdmin extends LeftAndMain implements PermissionProvider
+        $response->setBody(json_encode([
+            'files' => $items,
+            'count' => count($items),
+            'parent' => $folder->ParentID, // grandparent
```
This might need a check on $folder as well?
`$folder ? $folder->ParentID : 0`

tractorcow commented on code/Controller/AssetAdmin.php:
```diff
@@ -78,335 +75,405 @@ class AssetAdmin extends LeftAndMain implements PermissionProvider
+                'responseFormat' => 'json',
+            ],
+            'updateFileEndpoint' => [
+                'url' => Controller::join_links($baseLink, 'api/updateFile'),
```
I'm not sure I'm happy 100% with this; having an API is fine, but we've developed the form schema for this.

Is it going to be ok for me to migrate file editing to the form schema, once I've been able to complete that work? I would leave the other endpoints in place.

I'd probably not try to apply form schema to actions that are simply uploads or typing out new folder names, however.

chillu replied:
It's the existing approach. Will be fixed through https://github.com/silverstripe/silverstripe-asset-admin/issues/171

tractorcow commented on code/Controller/AssetAdmin.php:
```diff
@@ -78,335 +75,405 @@ class AssetAdmin extends LeftAndMain implements PermissionProvider
+     * @return SS_HTTPResponse
      */
-    public function getCurrentFolderID()
+    public function apiUpdateFile(SS_HTTPRequest $request)
```
Continuing from the comment I left above, I would prefer this action was wrapped in a Form instance, and invoked via the form schema.

tractorcow commented on code/Controller/AssetAdmin.php:
```diff
@@ -78,335 +75,405 @@ class AssetAdmin extends LeftAndMain implements PermissionProvider
+        return (new SS_HTTPResponse(json_encode($result)))->addHeader('Content-Type', 'application/json');
+    }
+
+    public function legacyRedirectForEditView($request)
```
PHPDoc please. :D

chillu replied:
Just moved this existing function further down in the code, but yep, I'll add some.

tractorcow commented on code/Controller/AssetAdmin.php:
```diff
@@ -78,335 +75,405 @@ class AssetAdmin extends LeftAndMain implements PermissionProvider
     }
 
-    public function addfolder($request)
+    public function apiCreateFolder(SS_HTTPRequest $request)
```
Each of these api* methods will need phpdoc.

tractorcow commented on code/Controller/AssetAdmin.php:
```diff
@@ -673,4 +580,182 @@ public function providePermissions()
+     * @param Folder
+     * @return Form
+     */
+    protected function getFolderEditForm(Folder $folder)
```
Note to self; Candidate for form schema-izeation, maybe?

chillu replied:
I'm not using this method yet anywhere BTW

tractorcow commented on code/Controller/AssetAdmin.php:
```diff
@@ -673,4 +580,182 @@ public function providePermissions()
+            ),
+            'title' => $file->Title,
+            'type' => $file->is_a('Folder') ? 'folder' : $file->FileType,
+            'category' => $file->is_a('Folder') ? 'folder' : $file->appCategory(),
```
Probably better to override these properties in Folder.

chillu replied:
I've migrated the code as-is. Do you want to raise this as a separate PR? It's a bit tricky to resolve, since the React code does checks like `item.type === 'folder'`. That's fine for this particular case since `folder` is hardcoded above, but `File->getFileType()` returns an abitrary, localised string which shouldn't be used for comparisons. So I could override `Folder->getFileType()` with `return 'folder'`, which seems inconsistent. Or I could use `return _t('Folder.SINGULARNAME')`, which we can't use for comparisons.

We shouldn't really use `Folder.ClassName` comparisons on the client, since you could create `Folder` subclasses. Does replacing `item.type === 'folder'` with `item.category === 'folder'` make sense? Or do we need a new property?

tractorcow commented on code/Controller/AssetAdmin.php:
```diff
@@ -673,4 +580,182 @@ public function providePermissions()
+            'owner' => null,
+            'parent' => null,
+            'attributes' => array(
+                'dimensions' => array(),
```
`$file->getDimensions('array')` ?

tractorcow commented on code/Controller/AssetAdmin.php:
```diff
@@ -673,4 +580,182 @@ public function providePermissions()
+        }
+
+        /** @var File $file */
+        if ($file->hasMethod('getWidth') && $file->hasMethod('getHeight')) {
```
File always has this method. Use `if($files->getIsImage())` instead.

chillu replied:
Same thing, that's existing code. But I'll fix it up.

tractorcow commented on client/src/state/gallery/GalleryActions.js:
```diff
@@ -9,7 +9,7 @@ export function setFolder(folderId) {
     dispatch({
       type: GALLERY.SET_FOLDER,
-      payload: { folderId },
+      payload: { ID: folderId },
```
Should we use `id` for lower camel case convention?

chillu replied:
It's a bit of a mess in there at the moment - I'm trying to use DataObject property naming as a first pass (`ParentID` rather than `folderId`), and where appropriate I'm using the uppercase convention. We'll need to go through the entire codebase for the lowercase work anyway, so there's not much point half doing it at the moment. This particular instance is just a side effect of making the API responses consistent.

tractorcow commented:
Not a bug, but a usability concern; I notice that `HTTP GET` methods are passing in the securityID. SecurityID is ignored in non-write actions, so it's misleading to have this passed in for such operations.

Having it included in all requests, including non-destruction actions, introduces an unnecessary appearance of reliance on this token.

As a general rule, if your action is destructive, it shouldn't be HTTP GET in the first place, and pass the CSRF via the request body / headers. Thus, I wouldn't except to see CSRF in any querystring, legacy code temporarily excused. :)

tractorcow commented on code/Controller/AssetAdmin.php:
```diff
@@ -78,335 +75,405 @@ class AssetAdmin extends LeftAndMain implements PermissionProvider
+
+        $folderID = (int)$params['id'];
+        if ($folderID === 0) {
+            $folder = new Folder();
```
Maybe better to leave this null and take up @flamerohr's suggestion. We don't need an empty object.

chillu commented:
Agreed that readonly requests shouldn't have CSRF, that was an oversight, applied it too broadly. Fixed now.

I think I've addressed all your feedback now, thanks for reviewing! Wanna check?
